### DISCUSSION TITLE
How to write "Display Name" and "Description" to Node

### DISCUSSION THREAD

**Discussion by Aposeltof** [gopcua/opcua#514] [Q&A] [answered]
First thanks a lot for this module! It's really great. 

For the moment we have used the browse and write value functions. Everything works fine.
The next step is to understand how to change a "display name" and "description" of an existing Node. 

Can you help me ?

I try it like that for the moment : 
```
func (s *instanceOPCClient) WriteDisplayNameByID(displayName string, ns uint16, id uint32) (success bool) {
	//Get Node id
	nid := ua.NewNumericNodeID(ns, id)

	//Open Node
	node := s.inst.uaClient.Node(nid)

	//Read current Node value to check it
	_, err := node.Value()
	if rslog.Check(err, slpOPC) {
		//Create a Variant from value
		variant, err := ua.NewVariant(displayName)
		if rslog.Check(err, slpOPC) {
			//Create a Write Request
			req := &ua.WriteRequest{
				NodesToWrite: []*ua.WriteValue{
					{
						NodeID:      nid,
						AttributeID: ua.AttributeIDDisplayName,
						Value: &ua.DataValue{
							EncodingMask: ua.DataValueValue,
							Value:        variant,
						},
					},
				},
			}

			//Write value to node
			_, err = s.inst.uaClient.Write(req)
			return rslog.Check(err, slpOPC)
		}
	}
	return false
}
```

**Comment by siscia** [comment]
Glancing at the code I don't see any obvious error.

I guess it didn't work, what error did you get?

**Comment by magiconair** [comment]
As a general Go comment: try to avoid nesting if statements and don't use helpers for error checking. It makes code harder to read since the idiomatic way of writing this is to fail fast. For example:

The other thing is that you should check the result code of the `Write` operation. If `err != nil` then that means that the `Write` operation has failed as a whole (e.g. server down, session closed, ...) but since you can update multiple nodes at the same time you need to check the `resp.Results` array for each node you've written.

```go 
func WriteDisplayNameByID(c *Client, displayName string, ns uint16, id uint32) error {
	// Get Node id
	nid := ua.NewNumericNodeID(ns, id)

	// Open Node
	node := c.Node(nid)

	// Read current Node value to check it
	_, err := node.Value()
	if err != nil {
		return err
	}

	// Create a Variant from value
	value, err := ua.NewVariant(displayName)
	if err != nil {
		return err
	}

	req := &ua.WriteRequest{
		NodesToWrite: []*ua.WriteValue{
			{
				NodeID:      nid,
				AttributeID: ua.AttributeIDDisplayName,
				Value: &ua.DataValue{
					EncodingMask: ua.DataValueValue,
					Value:        value,
				},
			},
		},
	}

	resp, err := c.Write(req)
	if err != nil {
		return err
	}

	if resp.Results[0] != ua.StatusOK {
		return resp.Results[0]
	}

	return nil
}
```

**Comment by magiconair** [comment]
As an aside: It might make sense to have a helper for setting attributes on the `Node` instance. Something like `n.SetAttribute(attrID ua.AttributeID, value interface{}) error`

**Reply by Aposeltof** [reply]
Yes it would be great

**Comment by Aposeltof** [comment]
Thanks for answering @magiconair and @siscia also for the general feedback.

I tried your code @magiconair . 

I get the next message error : 
       **_The attribute is not supported for the specified Node. StatusBadAttributeIDInvalid (0x80350000)_**

But it was because of the commented part.

Now I have the error 
**_User does not have permission to perform the requested operation. StatusBadUserAccessDenied (0x801F0000)_**

But I think it's more a node configuration error on the server side. 
I will try to correct it and try again.


```
func WriteDisplayNameByID(c *opcua.Client, displayName string, ns uint16, id uint32) error {
	// Get Node id
	nid := ua.NewNumericNodeID(ns, id)

	// Open Node
	//	node := c.Node(nid)

	// Read current Node value to check it
	//_, err := node.Value()
	//if err != nil {
	//		return err
	//}

	// Create a Variant from value
	value, err := ua.NewVariant(displayName)
	if err != nil {
		return err
	}

	req := &ua.WriteRequest{
		NodesToWrite: []*ua.WriteValue{
			{
				NodeID:      nid,
				AttributeID: ua.AttributeIDDisplayName,
				Value: &ua.DataValue{
					EncodingMask: ua.DataValueValue,
					Value:        value,
				},
			},
		},
	}

	resp, err := c.Write(req)
	if err != nil {
		return err
	}

	if resp.Results[0] != ua.StatusOK { 
		return resp.Results[0] //  // Here come the user access error
	}

	return nil
}
```

**Comment by Aposeltof** [comment, marked as answer]
Hello, it works now. 

**On the Client side in Go :** 
I read and modify the display name attribute which is a "Localized.Text". 
I convert it to a Variant then I can write it via the WriteRequest.

**Server side:**
I have to configure the "Write Mask" of the node so that the Display Name can be modified (Code 64).

**The code** 


```
func WriteDisplayNameByID(c *opcua.Client, displayName string, ns uint16, id uint32) error {
	// Get Node id
	nid := ua.NewNumericNodeID(ns, id)

	// Open Node
	node := c.Node(nid)

	// Read current DisplayName
	dName, err := node.DisplayName()
	if err != nil {
		return err
	}

	//Change text
	dName.Text = displayName

	// Create a Variant from dName (Struct LocalizedText)
	value, err := ua.NewVariant(dName)
	if err != nil {
		return err
	}

	req := &ua.WriteRequest{
		NodesToWrite: []*ua.WriteValue{
			{
				NodeID:      nid,
				AttributeID: ua.AttributeIDDisplayName,
				Value: &ua.DataValue{
					EncodingMask: ua.DataValueValue,
					Value:        value,
				},
			},
		},
	}

	resp, err := c.Write(req)
	if err != nil {
		return err
	}

	if resp.Results[0] != ua.StatusOK {
		return resp.Results[0]
	}

	return nil
}

```
